### PR TITLE
bquilain/hybridPMT merge to nuPRISM repo

### PR DESCRIPTION
**Pull request by pdeperio**
Work was done by @bquilain in https://github.com/bquilain/WCSim/pull/12. Please see that PR for discussion and merge modifications.

### CODE REVIEW

**Comment by bquilain** [comment]
This pull request contains the complete nuPRISM and hybrid merging done on my local fork. It basically contains:
-Possibility to have two types of PMT at construction.
-If several type of PMTs are used, have separated digitisers, dark rate, and triggers for each type.
-Radon generator is added.
-All float are converted to double.
-Very small change on the shape of the PMT holder in WCSimConstructPMT.cc.

**Comment by bquilain** [comment]
I have compared the hybrid and this merge branch results also. Everything is summarised here for HK, and it worked perfectly. It would be very nice if you could have a quick check on E61 side, as you would be much more able to detect inconsistencies than myself on that part
[HyperKSoftware_20210422_HyperKamiokandeSoftwareMeeting_MergingAndStatusOfWCSim_Quilain.pdf](https://github.com/nuPRISM/WCSim/files/6395094/HyperKSoftware_20210422_HyperKamiokandeSoftwareMeeting_MergingAndStatusOfWCSim_Quilain.pdf)

**Comment by bquilain** [comment]
Just for information, I have of course kept the mPMT geometry of the nuPRISM branch (except for the small difference I mentioned in the PMT holder). It slightly changed results for HK-hybrid, but I confirmed these geometry differences (mostly, the gel thickness) was the source. Therefore, HK will start using the updated design from next production.

**Comment by pdeperio** [comment]
> (except for the small difference I mentioned in the PMT holder)

Is this ^ referring to [this comment](https://github.com/bquilain/WCSim/pull/12#issuecomment-827512017
)? If so, do you actually mean you accepted the nuPRISM version and not "except"ed it?

File: macros/tuning_parameters.mac
Comment (akutsuR) [inline macros/tuning_parameters.mac]: Have we changed nominal value from 0.75 to 0.85?

File: src/WCSimConstructPMT.cc
Comment (akutsuR) [inline src/WCSimConstructPMT.cc]: We may want to add comment, in order to avoid confusion in future?

**Comment by bquilain** [comment]
> > (except for the small difference I mentioned in the PMT holder)
> 
> Is this ^ referring to [this comment](https://github.com/bquilain/WCSim/pull/12#issuecomment-827512017)? If so, do you actually mean you accepted the nuPRISM version and not "except"ed it?

Actually, I excepted it and used the hybrid one, which seemed to me more flexible. Could you check that part, and let me know if you disagree?

**Comment by pdeperio** [comment]
ok, then i think it's related to [this PR](https://github.com/nuPRISM/WCSim/pull/49); @nickwp ?

**Comment by nickwp** [comment]
Sorry I meant to comment on this before but have been a bit overwhelmed with other tasks.
Patrick is right that it's related to that PR. The mPMT geometry was updated significantly by a student who was doing some reflector optimisation. The version we kept matches the realistic mechanical design we have for WCTE and IWCD. For those experiments, we do want to keep the new design and not revert to the old one. But for HK I can see that it could be useful to have the flexibility of the old design... I'm not sure what the best solution is. (well the best solution I guess would be to add greater flexibility into the new design, I'm not sure how easy that will be though given the relatively significant changes)

File: include/WCSimWCPMT.hh
Comment (akutsuR) [inline include/WCSimWCPMT.hh]: Why don't we do something like `void SetRelativeDigitizedHitTime (bool val, G4double time=-1.) { RelativeHitTime = val; fFirst_Time = time; }`?

**Comment by bquilain** [comment]
Hi @nickwp . Thank you very much for your answer, it is quite clear. The HK mPMT geometry is so far, driven by the E61 one. Moreover, no one have planned or plans to do any study of the shape. So, though this is better to have some flexibility, it is not mandatory, especially if it comes at the cost of re-writing the new geometry correctly. So, I propose to revert to E61 one.

**Comment by bquilain** [comment]
> Hi @nickwp . Thank you very much for your answer, it is quite clear. The HK mPMT geometry is so far, driven by the E61 one. Moreover, no one have planned or plans to do any study of the shape. So, though this is better to have some flexibility, it is not mandatory, especially if it comes at the cost of re-writing the new geometry correctly. So, I propose to revert to E61 one.

Before doing this, I need to check impact on HK. Let me re-generate some studies before.

**Comment by bquilain** [comment]
Checks are done, and PMT holder change of shape does not change the HK results. It may only impact when we change the reflector shape too much, but this is not an issue within the reflector range of variation used so far.

**Comment by nickwp** [comment]
That's interesting that you don't see a significant difference for HK. When these changes were implemented we did see a reduction in light collection relative to the original version (by around 20%). If you can access the Canadian HK wiki, a summary is [here](https://www.hyperk.ca/mpmt/2020-meetings/2020-08-04/summary-so-far-reflector-optimization-i.pdf/view) where the "WCSim" design in that pdf is the original version and the "solidworks" design in the pdf is the new one, based on what is actually realistic to build.

**Comment by nickwp** [comment]
Ah, maybe you are only looking at the "PMT holder" specifically? I think unfortunately we do also need to use the new version of the reflector, too (and generally other changes to the mPMT construction in the nuPRISM version). Although we saw the significant reduction in light collection, the new version was still preferred since it is based on a realistic mechanical design possible to build, where the old model was not. I don't remember the details of this unfortunately, but can ask Mark Hartz about it.

**Comment by bquilain** [comment]
> That's interesting that you don't see a significant difference for HK. When these changes were implemented we did see a reduction in light collection relative to the original version (by around 20%). If you can access the Canadian HK wiki, a summary is [here](https://www.hyperk.ca/mpmt/2020-meetings/2020-08-04/summary-so-far-reflector-optimization-i.pdf/view) where the "WCSim" design in that pdf is the original version and the "solidworks" design in the pdf is the new one, based on what is actually realistic to build.

That's interesting. When you did these changes, did the only modified part was the PMT holder (PMTHolderZ, R etc.), or did you modified other parts also? Also, there is no reflector used in the current HK mPMT design (there is, but length is set to 0). Could it be the cause of it?

**Comment by bquilain** [comment]
> Ah, maybe you are only looking at the "PMT holder" specifically? I think unfortunately we do also need to use the new version of the reflector, too (and generally other changes to the mPMT construction in the nuPRISM version). Although we saw the significant reduction in light collection, the new version was still preferred since it is based on a realistic mechanical design possible to build, where the old model was not. I don't remember the details of this unfortunately, but can ask Mark Hartz about it.

If I did not make any mistake, all the other parts of the mPMT were updated, so there is no worry. I was referring here only to the PMT holder, as all other parts should have been updated in WCSimConstructPMT.cc. Of course, in WCSimDetectorConfig.cc, I still have set reflector to length = 0 for HK (not for E61) so far. This being said, checking I (or automatic merging) did not make any mistake would be very nice.

**Comment by nickwp** [comment]
Oh I see, thank you for the explanation! I'm sorry I didn't realise the other changes were also still there (should have checked the code before commenting).
In fact that all makes sense now and is actually an interesting finding - it shows that the ~20% reduction in light collection was really due to the reflector, as we expected, and not due to any other changes that came alongside it, since without the reflector in HK there is no similar reduction in light collection.

File: include/WCSimWCTrigger.hh
Comment (akutsuR) [inline include/WCSimWCTrigger.hh:335]: Are there any reason to change the default value? (20 is the current default value used in HK far detector side?)

File: src/WCSimConstructMaterials.cc
Comment (akutsuR) [inline src/WCSimConstructMaterials.cc]: Where did the difference come from?

File: src/WCSimTrackingAction.cc
Comment (akutsuR) [inline src/WCSimTrackingAction.cc:69]: Is this for simulating NEUT's particles?

File: src/WCSimTuningParameters.cc
Comment (akutsuR) [inline src/WCSimTuningParameters.cc]: Where does the difference come from?

File: macros/tuning_parameters.mac
Comment (bquilain) [reply macros/tuning_parameters.mac]: No, we have kept 0.75. In HK, we used the tuning_parameters.mac file in the main repo, not in the macros. You are totally right that such a conflicting situation should be avoided, so let me change this file also to use a scattering correction factor of 0.75

File: src/WCSimConstructPMT.cc
Comment (bquilain) [reply src/WCSimConstructPMT.cc]: If you agree, I even propose to remove it. I thought it would be useful for your checks, but think it can be removed and people can use old version if needed. Let me remove it.

File: macros/tuning_parameters.mac
Comment (bquilain) [reply macros/tuning_parameters.mac]: Done in the commit below

File: src/WCSimConstructPMT.cc
Comment (bquilain) [reply src/WCSimConstructPMT.cc]: Done in the commit below

File: include/WCSimWCPMT.hh
Comment (bquilain) [reply include/WCSimWCPMT.hh]: If you check the code WCSimWCPMT.cc, there is a check that fFirst_time = -1 to identify if we are in relative time or not. So, moving to your proposal would need a change in the code. I personally think that the code is fine as it is, but I am open to modify  it if needed. Please let me know if you see any need of changing it, and any potential application.

File: include/WCSimWCTrigger.hh
Comment (bquilain) [reply include/WCSimWCTrigger.hh:335]: Thank you for raising this point. Actually, this has been a modification I have done a long time ago to have a default low E trigger. I am happy to change it back, but I will update the mac file to set the threshold lower for HK.

File: src/WCSimTrackingAction.cc
Comment (bquilain) [reply src/WCSimTrackingAction.cc:69]: Good point. Frankly speaking, I do not know. I will check this part.

File: include/WCSimWCTrigger.hh
Comment (bquilain) [reply include/WCSimWCTrigger.hh:335]: Done

File: src/WCSimConstructMaterials.cc
Comment (bquilain) [reply src/WCSimConstructMaterials.cc]: Thank you for seeing it. I think it is a bug from a very old study I did 3 years ago, trying to increase blacksheet reflection by +10% hardcodedly, instead of using mac file. I have modified it back to 1.6, than you very much for seeing it. The difference was found to be negligible for E61 (so even more for HK), so I do not think we should re-check results.

File: src/WCSimTuningParameters.cc
Comment (bquilain) [reply src/WCSimTuningParameters.cc]: I do not know actually. Anyway, for all HK analyses, we used 1.30 in the mac file, so I have now changed it back to 1.30 as you suggest. All my thanks

File: include/WCSimPrimaryGeneratorAction.hh
Comment (bquilain) [inline include/WCSimPrimaryGeneratorAction.hh:49]: @akutsuR and @nickwp , do you need that GetMode returns and "InteractionType_t" for E61

**Review by bquilain — APPROVED** [review]

File: include/WCSimWCPMT.hh
Comment (akutsuR) [reply include/WCSimWCPMT.hh]: Thank you for the explanation. I understand the need.

File: include/WCSimPrimaryGeneratorAction.hh
Comment (akutsuR) [reply include/WCSimPrimaryGeneratorAction.hh:49]: @bquilain and @nickwp I have never used it for my IWCD analyses and the IWCD analysis codes used for the XSec analyses also does not use it. (We have used copy of NEUT output, which is produced in WCSim)

**Review by nickwp — COMMENTED** [review]
Thank you very much for the huge amount of work on this!
I'm sorry that my participation in this has been delayed by some other urgent studies at the moment, so I will need more time to review more carefully.
For now I have just a couple of comments:

I think a more careful merging of the makefile and cmake files should be done. For the current version, the build fails for me with various different errors when and either cmake or GNU make and root 6, or using GNU make and root 5. It only works with cmake and root 5 for me. @JacekHoleczek did some work on the WCTE fork to make the build process better and we have merged this into the nuPRISM fork. This includes making everything work for both root 5 and root 6, on various different systems, with both cmake and GNU make. Other changes were to ensure that all cmake generated files really go to the build directory and not the source directory, and other changes to make it more "robust". Some of these changes are reverted when combining with the other changes made by Guillaume also for root 6 compatibility. I will do some testing with both versions, to try to understand Guillaume's changes (which for me are not needed to work with root 6) and why the current version does not build on my system. Hopefully we can combine the two versions in a way that increases compatibility generally.

Another concern I have is regarding the change from floats to doubles. I found that after this change the file sizes did have a significant increase in size. On the the ComputeCanada cluster, disk space available to us is currently very limited (in fact right now it is already exceeded) and for the machine learning work we are simulating a lot of events, so WCSim files are the largest use of our storage space. The number of events needed for all our ML studies would be impossible to store with a significant increase in file size. I understand this change was necessary to precisely store times for the supernova events, where the times are offset by a larger value and so needs to be both large and precise, so a float doesn't work. I will do some testing and send a separate email with the people involved in the discussion in [this PR](https://github.com/WCSim/WCSim/pull/280), where the reason for the change was discussed, to discuss whether we can find another solution.

File: include/WCSimPrimaryGeneratorAction.hh
Comment (nickwp) [reply include/WCSimPrimaryGeneratorAction.hh:49]: I also have never used this. It looks like this change was introduced in [this commit](https://github.com/nuPRISM/WCSim/commit/ac4d7078c105e0949e48ff27775a9b70a6053677) by Tom Feusels in 2016, so it's been there for a long time and possibly some old analysis code uses it. Perhaps @tfeusels or @mscott201 would know?

Comment (nickwp) [reply include/WCSimPrimaryGeneratorAction.hh:49]: (I note that the commit comment when introducing InteractionType_t says that it "breaks backwards compatibility for Nuance, which is deprecated anyway" ... but five years later nuance input is still being used in some cases)

**Comment by akutsuR** [comment]
> Thank you very much for the huge amount of work on this!
> I'm sorry that my participation in this has been delayed by some other urgent studies at the moment, so I will need more time to review more carefully.
> For now I have just a couple of comments:
> 
> I think a more careful merging of the makefile and cmake files should be done. For the current version, the build fails for me with various different errors when and either cmake or GNU make and root 6, or using GNU make and root 5. It only works with cmake and root 5 for me. @JacekHoleczek did some work on the WCTE fork to make the build process better and we have merged this into the nuPRISM fork. This includes making everything work for both root 5 and root 6, on various different systems, with both cmake and GNU make. Other changes were to ensure that all cmake generated files really go to the build directory and not the source directory, and other changes to make it more "robust". Some of these changes are reverted when combining with the other changes made by Guillaume also for root 6 compatibility. I will do some testing with both versions, to try to understand Guillaume's changes (which for me are not needed to work with root 6) and why the current version does not build on my system. Hopefully we can combine the two versions in a way that increases compatibility generally.
> 
> Another concern I have is regarding the change from floats to doubles. I found that after this change the file sizes did have a significant increase in size. On the the ComputeCanada cluster, disk space available to us is currently very limited (in fact right now it is already exceeded) and for the machine learning work we are simulating a lot of events, so WCSim files are the largest use of our storage space. The number of events needed for all our ML studies would be impossible to store with a significant increase in file size. I understand this change was necessary to precisely store times for the supernova events, where the times are offset by a larger value and so needs to be both large and precise, so a float doesn't work. I will do some testing and send a separate email with the people involved in the discussion in [this PR](https://github.com/WCSim/WCSim/pull/280), where the reason for the change was discussed, to discuss whether we can find another solution.

@nickwp Thank you for the careful look at CMake and GNU files. I did not notice this as CMake with ROOT5 worked for me. I totally agree with the proposal of combing the two versions into a single version.

 As for the precision issue, the HK-LOWE people could adopt the MDT, so that they would be able to produce SN events or other LOWE specific events, which require the double precision for large time offset, using WCSim's output with the float precision for hit time, since I guess that the key point is to add a precise timing offset to PMT hits. But, I might have misunderstood their point.

**Comment by bquilain** [comment]
> Thank you very much for the huge amount of work on this!
> I'm sorry that my participation in this has been delayed by some other urgent studies at the moment, so I will need more time to review more carefully.
> For now I have just a couple of comments:
> 
> I think a more careful merging of the makefile and cmake files should be done. For the current version, the build fails for me with various different errors when and either cmake or GNU make and root 6, or using GNU make and root 5. It only works with cmake and root 5 for me. @JacekHoleczek did some work on the WCTE fork to make the build process better and we have merged this into the nuPRISM fork. This includes making everything work for both root 5 and root 6, on various different systems, with both cmake and GNU make. Other changes were to ensure that all cmake generated files really go to the build directory and not the source directory, and other changes to make it more "robust". Some of these changes are reverted when combining with the other changes made by Guillaume also for root 6 compatibility. I will do some testing with both versions, to try to understand Guillaume's changes (which for me are not needed to work with root 6) and why the current version does not build on my system. Hopefully we can combine the two versions in a way that increases compatibility generally.
> 
> Another concern I have is regarding the change from floats to doubles. I found that after this change the file sizes did have a significant increase in size. On the the ComputeCanada cluster, disk space available to us is currently very limited (in fact right now it is already exceeded) and for the machine learning work we are simulating a lot of events, so WCSim files are the largest use of our storage space. The number of events needed for all our ML studies would be impossible to store with a significant increase in file size. I understand this change was necessary to precisely store times for the supernova events, where the times are offset by a larger value and so needs to be both large and precise, so a float doesn't work. I will do some testing and send a separate email with the people involved in the discussion in [this PR](https://github.com/WCSim/WCSim/pull/280), where the reason for the change was discussed, to discuss whether we can find another solution.

Hi @nickwp and @akutsuR . Thank you very much for your feedback, and no problem if you are very busy with other items. Let me address your 2 points:
-I do agree with you for the cmake/makefile part. As for me, the cmake part worked perfectly for both ROOT5 and ROOT6 on 3 different clusters: sukap, French and local LLR one. So, it might need only little correction to perfectly work. As for the makefile building part, it has clearly much more issues. Generally, I think it would be ideal to work towards a joint method which increases compatibility. I am really not a makefile expert, but do you think that you would have time to take care about this? If your time is limited now but you think you would like to contribute, we could of course slightly delay the release. Also, I wished to ask @gpronost if he had some thoughts on this compatibility issue?
-For float or double: I agree we should seriously consider this issue. The storage issue for long-term HK far detector should be checked, but I believe a factor 2 is a minor problem compared to other uncertainties we have on it. As for the issue you have with ML on Canada's cluster, may I ask you if the storage issue is temporary, or do you expect to face this issue for a very long time? If this is temporary, we could think of a temporary patch/additional code (apply a code on WCSim output converting double to float) until your issue is solved.

**Comment by tdealtry** [comment]
Re: the file size, the way it was handled on WCSim/WCSim#280 was to only store times as `double`s in the output file, rather than everything (but still use `double` as the type for the set arguments, and still use `double` for everything internally in the running code)
On a [10 event 500 MeV electron SK geometry file-size test](https://github.com/WCSim/WCSim/pull/280#issuecomment-614597467), this gives a 20% increase in file size, instead of a 60% increase. Which is of course a non-negligible increase, but a lot better

**Comment by gpronost** [comment]
I will have a look at the compilation.
Regarding float or double, I think time needs to be double, other variable may be ok as float.

**Comment by gpronost** [comment]
I didn't have any issue with cmake compiling with ROOT5 or 6. 
For "make", I modified GNUmakefile_full in order to solve some bug (linked to the Root Dictionnary production), however, the fix is incomplete: we need to compile twice for the compilation to succeed, as G4 doesn't include the Root Dictionnary in the source list (since the Dictionnary doesn't exist when we compile for the first time)

**Comment by JacekHoleczek** [comment]
Maybe:
`all: rootcint movedict shared libWCSim.a lib bin`

**Comment by pdeperio** [comment]
@gpronost would you be able to review https://github.com/WCTE/WCSim/pull/1 and https://github.com/WCTE/WCSim/pull/3 if there are any conflicts/issues with your changes?

And vice versa, could you send the link of your PR for build system changes so @JacekHoleczek can review?

**Comment by JacekHoleczek** [comment]
Actually, I think my changes (building procedure related) were done in two PRs:

[WCTE#1](https://github.com/WCTE/WCSim/pull/1)

[WCTE#3](https://github.com/WCTE/WCSim/pull/3)

**Comment by gpronost** [comment]
> Maybe:
> `all: rootcint movedict shared libWCSim.a lib bin`
Thanks, yes, it works fine. 




> Actually, I think my changes (building procedure related) were done in two PRs:
> 
> [WCTE#1](https://github.com/WCTE/WCSim/pull/1)
> 
> [WCTE#3](https://github.com/WCTE/WCSim/pull/3)

I managed to compile with cmake ROOT6 with WCTE CMakeList.txt, but it's a bit weird: normally ROOT_GENERATE_DICTIONARY should not work with / in ROOT6+, in addition, there is a missing header. With ROOT5 it didn't work.
You need to add WCSimRootTools.cc/.hh in ROOT_GENERATE_DICTIONARY and in add_library(WCSimRoot SHARED ... ).

For make, you should add "clean::" in the GNUmakefile_full to remove the Dict files. Except this, I had no issue to compile with either ROOT5 or ROOT6.

**Comment by JacekHoleczek** [comment]
Currently, the [WCTE / WCSim ](https://github.com/WCTE/WCSim) and the [nuPRISM / WCSim ](https://github.com/nuPRISM/WCSim) say that they require the "known to work" Geant4 10.1.p03. I imagine the majority of users follow this requirement.
However, this PR says that it needs Geant4 10.2 or later and was tested with Geant4 10.3.p03.
Perhaps the HK side upgraded Geant4, and we should follow.
If something on the HK side required this upgrade, what potential implication for IWCD/WCTE might arise?

**Comment by bquilain** [comment]
> Currently, the [WCTE / WCSim ](https://github.com/WCTE/WCSim) and the [nuPRISM / WCSim ](https://github.com/nuPRISM/WCSim) say that they require the "known to work" Geant4 10.1.p03. I imagine the majority of users follow this requirement.
> However, this PR says that it needs Geant4 10.2 or later and was tested with Geant4 10.3.p03.
> Perhaps the HK side upgraded Geant4, and we should follow.
> If something on the HK side required this upgrade, what potential implication for IWCD/WCTE might arise?

Hi @JacekHoleczek . Sorry for my slow answer. This upgrade has been made by @gpronost . Actually, my understanding is that 4.10.2 is required to use the Rn generator (Guillaume can comment which function is needed exactly). I believe we would like to have this generator in WCTE/IWCD, so it may be good to also follow up this requirement.

**Comment by gpronost** [comment]
Hi, yes, the requirement is only if you want to simulate radioactive decays (ex: Rn generator), it could work with older versions of geant4, but there were some report of issues (I don't know the details of these issues, I didn't test it myself).

File: include/WCSimWCPMT.hh
Comment (gpronost) [reply include/WCSimWCPMT.hh]: Sorry, I was missing this message, actually, @akutsuR proposal is good since we have:

`   void SetRelativeDigitizedHitTime (bool val) { RelativeHitTime = val; fFirst_Time = -1; }
`
`   void SetRelativeDigitizedHitTime (bool val, G4double time) { RelativeHitTime = val; fFirst_Time = time; }
`
These fFirst_Time = -1 is used for 20" PMT, whereas fFirst_Time = time is used for mPMT (feeding them with the first time of the 20" PMT hits)

**Comment by bquilain** [comment]
@akutsuR and @gpronost Sorry for delay, but I just had time to work on a remaining issue shown by Akutsu-san: time shift of ~1ns between the digitized hits of this merged version and the previous E61 branch. I remarked that th fFirst_time, used normally for RelativeTime case, has a default value of -1. Consequently, this shift was also applied even in non Relative timing case. I set the default value to 0. @gpronost , could you let me know if you see any issue in doing so for the Rn generator? @akutsuR , could you let me know if the discrepancy in WCTE disappeared (I validated the shift for HK-FD sample, but your check would be really appreciated)? All my thanks

**Comment by akutsuR** [comment]
@

> @akutsuR and @gpronost Sorry for delay, but I just had time to work on a remaining issue shown by Akutsu-san: time shift of ~1ns between the digitized hits of this merged version and the previous E61 branch. I remarked that th fFirst_time, used normally for RelativeTime case, has a default value of -1. Consequently, this shift was also applied even in non Relative timing case. I set the default value to 0. @gpronost , could you let me know if you see any issue in doing so for the Rn generator? @akutsuR , could you let me know if the discrepancy in WCTE disappeared (I validated the shift for HK-FD sample, but your check would be really appreciated)? All my thanks

@bquilain, thank you very much for working on this and identifying the cause. I will generate additional MC to see if the shit has disappeared in the IWCD side. I will let you know the result as soon as I finish.

**Comment by gpronost** [comment]
Hi, I think '0' is not a good idea as it could actually be the true first time (in some rare cases). I think a flag would be better.

**Comment by akutsuR** [comment]
@bquilain I generated additional MC with your modification. The 1ns shift is not seen in the additional MC. (please see page 3 in [the slides](https://www.dropbox.com/s/1m1xiofrx7pgiub/validation_WCSimIWCD_PR61_v2.pdf?dl=0))

**Comment by gpronost** [comment]
I put a PR to use a boolean flag instead of FirstTime == 0 as condition to change the FirstTime value
https://github.com/bquilain/WCSim/pull/19

**Comment by bquilain** [comment]
> @bquilain I generated additional MC with your modification. The 1ns shift is not seen in the additional MC. (please see page 3 in [the slides](https://www.dropbox.com/s/1m1xiofrx7pgiub/validation_WCSimIWCD_PR61_v2.pdf?dl=0))

@akutsuR This is excellent news, thank you so much for this super prompt check!

**Comment by bquilain** [comment]
@akutsuR @pdeperio @gpronost @nickwp @JacekHoleczek Do you see any remaining issue which prevent from completing the pull request? Any additional checks you'd like to do?

**Comment by bquilain** [comment]
Sorry for bothering you again. I will be unavailable to work on the software, apart from debugging, from the end of this week (and for probably 1.5 month). It should not be a reason for rushing, but I would like to assess the level of checks you'd like to do before merging happen. Basically, if nothing major is forecast, I hope we can merge within this week so the software is ready for any users from now. It would allow @akutsuR to make his tool compatible, Tom to work on this branch, calibration studies for HK to be prepared using this branch and event display to be exported. @nickwp @akutsuR @pdeperio , could you let me know what would be your plan/timeline concerning this PR?

**Comment by nickwp** [comment]
Sorry for the delay replying. From my side ideally I would have liked to find time to fully review the changes, but I think given the time constraints now and the checks that @akutsuR has already done, I'm happy for this to be merged as it is. I had a quick look again now, and could not see any obvious problems. Thanks for the huge amount of work on it!

**Comment by bquilain** [comment]
> Sorry for the delay replying. From my side ideally I would have liked to find time to fully review the changes, but I think given the time constraints now and the checks that @akutsuR has already done, I'm happy for this to be merged as it is. I had a quick look again now, and could not see any obvious problems. Thanks for the huge amount of work on it!

Hi @nickwp . First, I am very sorry that the PR merge has been tight, basically due to my own agenda. I thought about doing a release during the summer, which could be nice for everyone to use. May I ask you what would be your schedule for doing the checks you'd like, as I would be happy to wait for it before doing the release.

**Comment by pdeperio** [comment]
@nickwp should we tag `develop` now before merging any other PRs?

also, how should we make use of the `master` branch?

**Comment by nickwp** [comment]
> @nickwp should we tag `develop` now before merging any other PRs?

Seems like a good idea, to tag the version that was tested by @akutsuR and presented at the collaboration meeting, which I think is still the current version?

> also, how should we make use of the `master` branch?

I'm not sure how these two branches were used in the past? I think I always used `develop` because `master` seemed outdated. Maybe now is a good time to merge everything from `develop` into `master`, then continue working on develop until the next major version? Unless there was already some scheme for how to use the two branches?

**Comment by pdeperio** [comment]
@mscott201  ^ ?

**Comment by mscott201** [comment]
The develop and master branches were supposed to just follow the main WCSim develop and master - this was to allow us to include updates from the main WCSim but also have our own branches (nuPRISM/develop etc.) with IWCD specific code.

For nuPRISM/develop and nuPRISM/master, master was the 'freeze' branch, according to the gitflow workflow.  Any merge from develop to master should be tagged as a release version of the IWCD code.  I did this for production versions, which is probably why it hasn't changed for a long time.

**Pull request merged**
